test/slipstream/connection_test.exs

```diff
@@ -179,4 +179,73 @@ defmodule Slipstream.ConnectionTest do
       refute c.socket |> await_disconnect!() |> connected?
     end
   end
+
+  describe "given an open connection socket with heartbeat interval set very low" do
+    setup c do
+      conn = self()
+      stream_ref = make_ref()
+
+      @gun
+      |> stub(:open, fn _host, _port, _opts ->
+        # N.B.: we're providing the test process as the conn so it makes life
+        # easier later when we expect messages and want to send them to the
+        # test process for an assert_receive/2
+        {:ok, conn}
+      end)
+      |> stub(:ws_upgrade, fn _conn, _path, _headers, _opts ->
+        send(
+          self(),
+          {:gun_upgrade, conn, stream_ref, ["websocket"], _headers = []}
+        )
+
+        stream_ref
+      end)
+
+      config =
+        c.config
+        # 20ms between heartbeat requests
+        |> Keyword.put(:heartbeat_interval_msec, 20)
+
+      [config: config, conn: conn, stream_ref: stream_ref]
+    end
+
+    test """
+         when we snub heartbeat responses to the connection process,
+         then the client will disconnect
+         """,
+         c do
+      conn = c.conn
+
+      @gun
+      |> expect(:ws_send, 1, fn ^conn, {:text, heartbeat_request} ->
+        request =
+          heartbeat_request
+          |> IO.iodata_to_binary()
+          |> Jason.decode!()
+
+        send(conn, {:request, request})
+
+        :ok
+      end)
+      |> expect(:close, 1, fn ^conn -> :ok end)
+
+      socket = c.config |> connect!() |> await_connect!()
```

Comment:
the other tests in this file use Slipstream's synchronous API (the `await_*` family of functions) to test various things about the nitty-gritty details of a connection and how connections interoperate with `:gun`

with the synchronous API, the `self()` current process _is_ the client (think of scripting websocket stuff instead of writing a robust async, gen-server-style module)